source/slang/core.meta.slang

```diff
@@ -480,6 +475,16 @@ ${{{{
     }
 }}}}
 {
+
+    typedef $(kBaseTypes[tt].name) ThisType;
+
+    // Declare initializer of `This` type for default initialization
+    [ForceInline]
+    __init()
+    { 
+        this = __default<ThisType>();
+    }
```

Comment:
I see that cpp files checks the option like this,
````
this->getOptionSet().getBoolOption(CompilerOptionName::ZeroInitialize)
```
and change the behavior based on the command-line argument.
But this line doesn't check the option and it will always initialize the variables with zero, isn't it?
If it is the case, I am worried that Slang generated code will pay the cost of initialization even when they shouldn't be initialized.

If my understanding is incorrect, please let me know how it works.

Reply:
stdlib structs do not set init expressions if they have an empty `initExpr`.
This is due to the following code having `!isFromStdLib` to ensure we don't default init `(.*).meta.slang` types unless `CompilerOptionName::ZeroInitialize` is enabled.

`SemanticsDeclBodyVisitor::checkVarDeclCommon`
`else if(!varDeclBaseType || !varDeclBaseType->getDeclRef() || !isFromStdLib(varDeclBaseType->getDeclRef().getDecl()))`

I don't think this affects raytracing types, but it seems it does not (falcor works)

Reply:
I don't think I am following your explanation but it raised another question.
When slangc runs for the first time, it compiles stdlib and it is cached. And next time we run slangc, we don't recompile stdlib. 

From this point, the run time compiler option will have no impact on how stdlib gets compiled, because the compilation doesn't happen anymore.

Is the implementation consider this case?

Reply:
The part I don't understand is that you said, 
```
stdlib structs do not set init expressions if they have an empty initExpr.
```
But it looks like you are setting initExpr here.
```
this = __default<ThisType>();
```
I think "__default<ThisType>()" is initExpr, isn't it?

Reply:
> Does the implementation consider this case?

This has 2 parts:
1. does IDefaultInitialize get added to meta.slang

stdlib will not automatically add `IDefaultInitialize` if compiled without.
This cannot be worked around since a user does not have the ability to compile (.).meta.slang

2. does it still work as expected

Yes. To simplify the answer I will use vec2 as an example: if we write `vec2 myVar` with the `-zero-initialize flag`, the following code will compile `vec2 myVar = vec2()`.

Reply:
`int a = __default<ThisType>();` => `int a = 0;`
`float a = __default<ThisType>();` => `float a = 0.0f;`

Reply:
So it will initialized to zero even when `-zero-initialize` is not set, is it the case?

Reply:
> So it will initialized to zero even when -zero-initialize is not set, is it the case?

no, the init is not automatically called. It just exists for us to use.
Image if I make an `interface thing{ __init(); }` --> this interface would not work with `int`, which is incorrect.

Reply:
After talking on Slack for a bit, it turned out that @ArielG-NV  was on an impression that we shouldn't be calling the constructors if the struc/class is defined in stdlib.

He thought that it is our design choice because he misinterpreted what Yong said on [one of comments](https://github.com/shader-slang/slang/pull/3987#pullrequestreview-2032213162),
> We probably want to exclude stdlib from this behaivor, though.

I don't believe that is what we want at all but I like to hear from @csyonghe if my understanding is correct.